database/mysql/mysql.go

```diff
@@ -252,65 +252,62 @@ func (m *Mysql) Close() error {
 }
 
 func (m *Mysql) Lock() error {
+	if !m.isLocked.CAS(false, true) {
+		return database.ErrLocked
+	}
+
 	if m.config.NoLock {
-		if !m.isLocked.CAS(false, true) {
-			return database.ErrLocked
-		}
 		return nil
 	}
 
 	aid, err := database.GenerateAdvisoryLockId(
 		fmt.Sprintf("%s:%s", m.config.DatabaseName, m.config.MigrationsTable))
 	if err != nil {
+		m.isLocked.Store(false)
 		return err
 	}
 
 	query := "SELECT GET_LOCK(?, 10)"
 	var success bool
 	if err := m.conn.QueryRowContext(context.Background(), query, aid).Scan(&success); err != nil {
+		m.isLocked.Store(false)
 		return &database.Error{OrigErr: err, Err: "try lock failed", Query: []byte(query)}
 	}
 
-	if success {
-		if !m.isLocked.CAS(false, true) {
-			return database.ErrLocked
-		}
-		return nil
+	if !success {
+		m.isLocked.Store(false)
```

Comment:
~Add a comment here about how removing the local lock in this situation is desired since the local lock is an optimization and it's not trying to reflect the state of the db lock.~

Actually, let's add our own CAS wrapper to automatically restore the lock state on error (in `database/util.go`) so we don't need to remember to do it:
```go
func casRestoreOnErr(lock *atomic.Bool, o, n bool, f func() error) error {
    if !lock.CAS(o, n) {
        return ErrLocked
    }
    if err := f(); err != nil {
        // Automatically unlock
        lock.Store(o)
        return err
    }
    return nil
}
```
Also, can you add tests for this wrapper?

Reply:
Thank you. Of course I will do it

Reply:
Done, review pls